### ISSUE TITLE
[Feature Request] CriPO: criterion-level credit assignment for rubric-based RL

### ISSUE THREAD

**Issue by Mantissagithub** [PrimeIntellect-ai/prime-rl#3278]
### Summary

[CriPO: Enhancing Rubric-based RL via Self-Distillation](https://arxiv.org/abs/2607.18082) (Xia et al.) targets two GRPO failure modes on rubric-graded tasks, both of which `grpo` here has by construction, since it collapses the rubric to one scalar (`rollout.reward`) and broadcasts one group-relative advantage over every token:

- **unexplored criteria**: no rollout in the group satisfies criterion `c_j`, so nothing points toward it.
- **suppressed criteria**: some rollout does satisfy `c_j`, but its aggregate advantage is negative (or zero while `|S_j| < G/2`), so the correct span is trained away with the rest of a response that scored badly elsewhere.

`L_CriPO = L_GRPO + β·L_OPSD`, β = 0.05/0.03 for Qwen3-1.7B/4B with 50-step linear warmup. Reported 62.4 vs GRPO's 59.2 on 1.7B, 69.6 vs 68.2 on 4B. Both branches use the live policy as its own criterion-conditioned teacher, the same trick `opsd` already implements here.

(shameless plug: I wrote up the method in detail [here](https://www.pradheep.dev/reading/cripo).)

I'd like to contribute this, but the two branches land very differently against the current architecture.

### Branch 1: advantage flipping, fits as-is

A counterfactual teacher pass asks the model to delete the satisfying behavior; tokens that drop under that instruction get `τ_flip = 0.1`, everything else keeps `A_i`:

```
T_i^s = { t | Δ_{i,t} > 0  ∧  p^T_{i,t}(y_{i,t}) < α · max_v p^T_{i,t}(v) },   α = 0.1
```

No wire change needed: `score_group` already receives the cohort before filtering (suppression needs sibling advantages), `rollout.rewards` already carries per-component values and weights for the top-K selection, and `assign_advantages` already takes a full-length per-token list.

One gap: the flip test needs `max_v p^T(v)` as well as `p^T(y_t)`, but `prefill_logprobs` requests `prompt_logprobs: 1` and flattens each entry to one float. Looks like a small client-side extension.

### Branch 2: behavior injection, needs a wire-format decision

Forward KL from a criterion-conditioned view to the plain view, on the best rollout, restricted to the smallest token set covering 95% of total divergence:

```
L_OPSD = 1/|T^u| · Σ_{t∈T^u} D_KL( sg[π_θ(·|x, C_u, y*_{<t})] ‖ π_θ(·|x, y*_{<t}) )
```

`ref_kl_loss_fn` can't express this. It computes *reverse* KL from one scalar per token:

```python
ref_kl = ref_logprobs - trainer_logprobs
pg_loss = keep_mask * ref_kl.detach() * importance_ratio
```

Forward KL's gradient is a teacher-weighted cross-entropy, so it needs `p^T` over the vocabulary, and `TrainingSample.ref_logprobs` is `list[float]`. This would mean a per-token top-k teacher stream, matching `prepare_batch` packing changes, and a fourth loss component.

Worth noting the paper's own ablation: standalone OPSD collapses, all five benchmarks below base. The value is specifically in localizing to the 95% set (34.6% of tokens in their first batch) on top of a GRPO backbone, so a dense approximation that dodges the wire change isn't a substitute.

### Questions

1. Is CriPO in scope? It's most useful to envs with per-criterion rewards rather than a single verifier score.
2. For branch 2, would you accept a top-k teacher stream on `TrainingSample` plus a forward-KL component, or is per-token top-k on the wire a cost you'd rather not take?
3. Two algorithms (`cripo_s` first, standalone) or one class with both branches behind flags? The paper evaluates them separately and `CriPO-S` alone already beats GRPO on 4B (68.9 vs 68.2).
4. CriPO assumes per-prompt dynamic rubrics, while reward components here are typically fixed per env. Should criterion text come from `rollout.info` or `task.data`? `opsd` sets a precedent with its configurable `demo_key`.

Happy to open branch 1 as a self-contained PR right away, and hold branch 2 until there's a call on question 2.

**Comment by Mantissagithub** [comment]
cc @hallerite @mikasenghaas

**Comment by RitwijParmar** [comment]
I put together the first branch as an opt-in `cripo_s` algorithm in [#3640](https://github.com/PrimeIntellect-ai/prime-rl/pull/3640). It keeps named verifier rewards separate and adds the counterfactual token bonus without changing the trainer wire format. I left the forward-KL branch out for now since that needs a top-k stream design.
